wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1425,3 +1392,288 @@ let construct_pointer_constraint (l_orig : Constr.z3_expr list) (env1 : Env.t)
+   - Looking at the target node for a backjump
+   - If the node has been visited more than [num_unroll] times, use the [loop_exit_pre] precondition
+   - Otherwise, decrement the [depth] map which tracks the unrollings for that node, and
+     recursively call [Precondition.visit_graph]. Because this function is defined in another
+     (later) module, we use open recursion via the [wp_rec_call] function reference. *)
+let loop_unroll (num_unroll : int)
+  : Env.t -> Constr.t -> start:Graphs.Ir.Node.t -> Graphs.Ir.t -> Env.t =
+  let module Node = Graphs.Ir.Node in
+  let find_depth depth node = Env.Unroll_depth.find depth (Node.label node)
+                              |> Option.value ~default:num_unroll
+  in
+  let decr_depth depth node =
+    Env.Unroll_depth.update depth (Node.label node)
+      ~f:(function None -> num_unroll - 1 | Some n -> n - 1)
+  in
+  let unroll env pre ~start:node g =
+    let tid = node |> Node.label |> Term.tid in
+    debug "Unrolling loop for %s%!" (Tid.to_string tid);
+    let depth = Env.get_unroll_depth env in
+    if find_depth depth node <= 0 then
+      let pre =
+        List.find_map [loop_exit] ~f:(fun spec ->
+            let* exit = spec node g in
+            let tid = exit |> Node.label |> Term.tid in
+            Env.get_precondition env tid)
+      in
+      let pre = match pre with
+        | Some p -> p
+        | None ->
+          warning "Trivial precondition is being used for node %s%!" (Tid.to_string tid);
+          let ctx = Env.get_context env in
+          Constr.trivial ctx
+      in
+      Env.add_precond env tid pre
+    else
+      begin
+        let updated_depth = decr_depth depth node in
+        let updated_env = Env.set_unroll_depth env updated_depth in
+        let _, env = visit_graph updated_env pre
+            ~start:node ~skip_node:(unreachable_from_start g node) g in
+        env
+      end
+  in
+  unroll
+
+(* Gets the target destination of a jump. We do not handle indirect jumps. *)
+let target (jmp : Jmp.t) : Tid.t option =
+  match Jmp.kind jmp with
+  | Goto (Direct tid) -> Some tid
+  | Int (_, tid) -> Some tid
+  | Call t -> begin
+      match Call.return t with
+      | Some (Direct tid) -> Some tid
+      | Some (Indirect _) | None -> None
+    end
+  | Goto (Indirect _) | Ret _ -> None
+
+(* Checks if [tid] corresponds to the tid of a block in a loop. *)
+let in_loop (loop : Graphs.Ir.Node.t group) (tid : Tid.t) : bool =
+  Seq.exists (Group.enum loop) ~f:(fun n ->
+      Tid.equal (n |> Graphs.Ir.Node.label |> Term.tid) tid)
+
+(* Assumes the loop header branches into two blocks. *)
+let get_branch (blk : Blk.t) : (Jmp.t * Jmp.t) option =
+  match Seq.to_list (Term.enum jmp_t blk) with
+  | [jmp1; jmp2] -> Some (jmp1, jmp2)
+  | _ -> None
+
+(* Gets the BAP expression that represents the condition for going into the
+   loop. *)
+let get_branch_condition (loop : Graphs.Ir.Node.t group) (jmp1 : Jmp.t)
```

Comment:
Still a bit confused if we're handling this correctly: we need to apply whatever's in this block to the condition as well, right?

Reply:
If that's the case, then I feel like a clever version of `visit_blk` should be used instead of this function.